### PR TITLE
fix(sphere-stock-import): Security vulnerability detected in hoek

### LINKED ISSUE #76: Several security vulnerabilities in sphere-stock-import detected
#### Description
Security vulnerability in sphere-stock-import detected in dependencies.

#### Expected Behaviour
No security vulnerability in dependencies

#### Current Behaviour
Security vulnerabilities found are
```
54 vulnerabilities (18 low, 19 moderate, 16 high, 1 critical)
```
#### Context
Sphere-stock-import package.json

#### Possible Solution
Updating of necessary dependencies to fix vulnerabilities

### PR DESCRIPTION
**Pull request by Babanila**
#### Summary
Security vulnerability detected in hoek

#### Description
Known moderate severity security vulnerability detected in hoek < 4.2.1 defined in package-lock.json.

Closes #76

### CODE REVIEW

File: src/coffee/stockimport.coffee [hunk 1]
```diff
@@ -383,8 +383,7 @@ class StockImport
                 debug "Error on handling 409 stock update error. Details: #{JSON.stringify(err)}"
                 Promise.reject err
           else
-            debug "Failed to retry the task after #{@max409Retries} attempts for stock #{JSON.stringify(entry)}"
-            Promise.reject err
+            Promise.reject new Error("Failed to retry the task after #{@max409Retries} attempts for stock #{JSON.stringify(entry)}")
```
Comment (daern91) [inline src/coffee/stockimport.coffee]: I dont think this should be part of this PR

Comment (butenkor) [reply src/coffee/stockimport.coffee]: Also please do not swallow original error objects as they contain a lot of valuable informations. See my comment [here](https://github.com/sphereio/sphere-stock-import/pull/75#discussion_r193980177)

Comment (butenkor) [reply src/coffee/stockimport.coffee]: We need here both the original error with its stack trace as also its context/custom message regarding amount of retries.

**Comment by lojzatran** [comment]
Hi @Babanila changing dependencies in package.json won't change nested dependencies. You need to do it like this: https://stackoverflow.com/questions/15806152/how-do-i-override-nested-npm-dependency-versions

**Comment by LEQADA** [comment]
When you add `hoek` as a dependency to your project `npm` will still download the vulnerable version too because it is used by some other dependencies.
We can force `npm` to use the version we want as @lojzatran says but this is mostly what we avoid to do. 
The right way is to update dependencies, if there is no fix for some dependency - check is the library still alive or not. If it's alive - maybe create a PR for that library (we live in an open source world and using it benefits). If the library is dead (no one updates it) then maybe we should replace it with something else.
In this case I can see that the highest vulnerable library in dependency chain is `request` library (`request > hawk > boom > hoek`).
The latest version of `request` library [doesn't use hawk](https://github.com/request/request/pull/2943). It means there is no `hoek` dependency at all.
So maybe we can just update the `request` library in our dependencies?

File: package.json [hunk 2]
```diff
@@ -42,31 +42,37 @@
   },
```
Comment (LEQADA) [inline package.json]: All you need to update in this file in order to get rid of `hoek` vulnerability is in this patch:
```
diff --git a/package.json b/package.json
index 83ebba7..baf4ac1 100644
--- a/package.json
+++ b/package.json
@@ -46,7 +46,7 @@
     "csv": "^1.1.0",
     "debug": "^2.2.0",
     "optimist": "^0.6.1",
-    "sphere-node-sdk": "^1.22.0",
+    "sphere-node-sdk": "^2.0.0",
     "sphere-node-utils": "^0.9.1",
     "tmp": "0.0.33",
     "underscore": "^1.8.3",
@@ -54,7 +54,7 @@
   },
   "devDependencies": {
     "coffee-script": "^1.12.7",
-    "coveralls": "2.11.2",
+    "coveralls": "2.13.2",
     "grunt": "0.4.5",
     "grunt-bump": "0.0.13",
     "grunt-cli": "0.1.13",
```

Comment (LEQADA) [reply package.json]: everything else should be reverted since they are not related to the issue context

**Comment by LEQADA** [comment]
But IMO, we should close this PR and start a new one which fixes all 48 vulnerabilities in this library

File: package.json [hunk 3]
```diff
@@ -45,16 +45,18 @@
     "bunyan-logentries": "^0.1.0",
     "csv": "^1.1.0",
     "debug": "^2.2.0",
+    "hoek": "^5.0.3",
```
Comment (daern91) [inline package.json]: @Babanila and @LEQADA is it really needed to add `hoek` as a dependency here? As far as I understood @LEQADA's comment we just needed to update `request` and possibly `coveralls` and `sphere-node-sdk`?

Comment (LEQADA) [reply package.json]: No, it's not needed. As well as adding `request` package as a dependency. 
And as I previously said it looks redundant to me to create a PR for each 48 vulnerabilities.

**Pull request closed without merging**